### PR TITLE
Allow restricted mode; include permission boundary option

### PR DESCRIPTION
**Pull request by lttmtins**

### CODE REVIEW

**Comment by lttmtins** [comment]
I actually don't know why this PR does not lead to the invocation of the github workflow. If I do the same PR to the fork's master branch, all of the fmt/tfsec checks are passing: https://github.com/sbpdvb/terraform-aws-mcaf-transfer/pull/1

File: README.md [hunk 1]
```diff
@@ -9,7 +9,9 @@ module "example-transfer" {
   name             = "example"
   endpoint_type    = "VPC"
   tags             = {}
-
+  restricted_mode = false
```
Comment (marwinbaumannsbp) [inline README.md]: can you format the readme?

File: variables.tf [hunk 2]
```diff
@@ -23,6 +23,11 @@ variable "users" {
   description = "A map with transfer users and configuration details"
 }
 
+variable "s3_id" {
```
Comment (marwinbaumannsbp) [inline variables.tf]: Can you add descriptions to the variables?

Comment (marwinbaumannsbp) [reply variables.tf]: Can you sort the variables? e.g. s3_id is before tags.

File: main.tf [hunk 3]
```diff
@@ -83,10 +86,19 @@ resource "aws_iam_role_policy" "user" {
...
-  home_directory = each.value.home_directory
-  role           = aws_iam_role.user[each.key].arn
-  server_id      = aws_transfer_server.default.id
+  user_name           = each.key
+  home_directory      = var.restricted_mode ? null : each.value.home_directory
+  home_directory_type = var.restricted_mode ? "LOGICAL" : "PATH"
+  role                = aws_iam_role.user[each.key].arn
+  server_id           = aws_transfer_server.default.id
+  tags                = var.tags
+  dynamic "home_directory_mappings" {
```
Comment (marwinbaumannsbp) [inline main.tf:97]: please add a blank line between tags and dynamic: https://www.terraform.io/language/syntax/style

File: variables.tf [hunk 4]
```diff
@@ -32,3 +37,23 @@ variable "vpc_endpoint" {
   default     = null
   description = "Optional VPC endpoint settings for your SFTP server"
 }
+variable "restricted_mode" {
```
Comment (marwinbaumannsbp) [inline variables.tf]: You are looping over var.users but are also using restrict_mode. Should restrict_mode not be part of the users variable? Same for the s3_id var?

File: variables.tf [hunk 5]
```diff
@@ -32,3 +37,23 @@ variable "vpc_endpoint" {
...
+  default = false
+}
+
+variable "permissions_boundary" {
+  type        = string
+  default     = null
+  description = "The permissions boundary to set on the role"
+}
+
+variable "login_banner" {
```
Comment (marwinbaumannsbp) [inline variables.tf]: This variable is not used

File: variables.tf (hunk 4 above)
Comment (lttmtins) [reply variables.tf]: The idea here is that with a single flag, you can switch between the original mode (for backwards compatibility), as well as towards a 'chroot' like style config. This makes the user config less error prone.

File: variables.tf (hunk 5 above)
Comment (lttmtins) [reply variables.tf]: it's removed. Will be added later.

File: variables.tf (hunk 2 above)
Comment (lttmtins) [reply variables.tf]: done

File: variables.tf (hunk 5 above)
Comment (lttmtins) [reply variables.tf]: removed.

File: main.tf [hunk 6]
```diff
@@ -83,10 +86,20 @@ resource "aws_iam_role_policy" "user" {
...
+  home_directory_type = var.restricted_mode ? "LOGICAL" : "PATH"
+  role                = aws_iam_role.user[each.key].arn
+  server_id           = aws_transfer_server.default.id
+  tags                = var.tags
+
+  dynamic "home_directory_mappings" {
+    for_each = var.restricted_mode ? [1] : []
+    content {
+      entry  = "/"
+      target = "/${var.s3_id}/${each.key}"
```
Comment (marwinbaumannsbp) [inline main.tf]: https://docs.aws.amazon.com/transfer/latest/userguide/logical-dir-mappings.html i think you need to add `transfer:${each.key}`

File: main.tf [hunk 7]
```diff
@@ -52,6 +53,7 @@ resource "aws_transfer_server" "default" {
   identity_provider_type = "SERVICE_MANAGED"
   logging_role           = aws_iam_role.server.arn
   tags                   = var.tags
+  security_policy_name   = var.transfer_security_policy
```
Comment (marwinbaumannsbp) [inline main.tf]: can you move security_policy_name above tags?

File: main.tf (hunk 6 above)
Comment (lttmtins) [reply main.tf]: I tested this before, and using ${transfer:UserName} as a var does not work in combination with Terraform. It accepts the var, but the AWS transfer service stores this as the actual username, not the string representing the var. This breaks the idempotency on the next terraform run. 
Doing it as above, the target is translated by Terraform to the actual value, and not the internal AWS transfer configuration engine.

File: variables.tf (hunk 4 above)
Comment (lttmtins) [reply variables.tf]: Agreed that it makes sense to make this a generic flag instead of a user-specific flag.

File: main.tf (hunk 7 above)
Comment (lttmtins) [reply main.tf]: Done

File: main.tf [hunk 8]
```diff
@@ -48,10 +49,12 @@ resource "aws_iam_role_policy_attachment" "server" {
...
 
 resource "aws_transfer_server" "default" {
-  endpoint_type          = var.vpc_endpoint != null ? "VPC" : "PUBLIC"
-  identity_provider_type = "SERVICE_MANAGED"
-  logging_role           = aws_iam_role.server.arn
-  tags                   = var.tags
+  endpoint_type                   = var.vpc_endpoint != null ? "VPC" : "PUBLIC"
+  identity_provider_type          = "SERVICE_MANAGED"
+  logging_role                    = aws_iam_role.server.arn
+  pre_authentication_login_banner = var.pre_login_banner != "" ? var.pre_login_banner : null
```
Comment (marwinbaumannsbp) [inline main.tf]: var.pre_login_banner is already default null, this if statement is not needed

File: variables.tf [hunk 9]
```diff
@@ -9,11 +9,34 @@ variable "name" {
...
+  default     = false
+  description = "Optional restricted mode."
+}
+
 variable "tags" {
   type        = map(string)
   description = "A mapping of tags to assign to the resources"
 }
 
+variable "transfer_security_policy" {
```
Comment (marwinbaumannsbp) [inline variables.tf:35]: Can you add a description?

File: variables.tf [hunk 10]
```diff
@@ -9,11 +9,34 @@ variable "name" {
...
+variable "pre_login_banner" {
+  type        = string
+  default     = null
+  description = "Login banner presented before logging on to the AWS Transfer server"
+}
+
+variable "restricted_mode" {
+  type        = bool
+  default     = false
+  description = "Optional restricted mode."
```
Comment (marwinbaumannsbp) [inline variables.tf]:
```suggestion
  description = "Restricts users so they only have access to their home directories"
```

File: README.md [hunk 11]
```diff
@@ -8,20 +8,22 @@ module "example-transfer" {
...
   name             = "example"
   endpoint_type    = "VPC"
-  tags             = {}
-
   endpoint_details = {
     address_allocation_ids = ["eipalloc-12345", "eipalloc-67890"]
     subnet_ids             = ["subnet-12345", "subnet-67890"]
     vpc_id                 = "vpc-123456"
   }
+  restricted_mode = false
```
Comment (marwinbaumannsbp) [inline README.md]: Can you modify the Readme example according to the styleguide?:

    meta-argument: `count` or `for_each`
    meta-argument: `providers`
    <empty line>
    source
    version
    [alphabetically sorted other fields]
    <empty line>
    blocks (each must be separated by empty lines)
    <empty line>
    meta-argument blocks: (`depends_on`, `lifecycle`, etc)

File: versions.tf [hunk 12]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">= 4.26.0"
```
Comment (marwinbaumannsbp) [inline versions.tf]: This looks too restrictive, I don't think the minimum required provider version needed for this module is 4.26.0. I see that the pre_authentication_banner has been added in 4.6.0.
```suggestion
      version = ">= 4.6.0"
```

File: variables.tf (hunk 10 above)
Comment (lttmtins) [reply variables.tf]: fixed

File: README.md (hunk 11 above)
Comment (lttmtins) [reply README.md]: done

**Review by marwinbaumannsbp — APPROVED** [review]

**Pull request merged**